File: frontend/src/types/campaign.ts
export const taskTypes = ["Brainstorm", "Copywriting", "Visual Asset"] as const;
export type TaskType = (typeof taskTypes)[number];

export const agentToolNames = [
  "quality_gate",
  "extract_brief",
  "brainstorm_angles",
  "write_copy",
  "generate_visual",
  "review_outputs",
] as const;
export type AgentToolName = (typeof agentToolNames)[number];

export type ToolCallStatus = "running" | "completed" | "failed";

export interface ToolCall {
  tool_run_id: string;
  tool: AgentToolName;
  status: ToolCallStatus;
  output_ref?: string | null;
  started_at: string;
  completed_at?: string | null;
}

export interface CampaignMessage {
  role: "user" | "assistant";
  content: string;
  created_at: string;
}

export type CampaignStatus =
  | "draft"
  | "queued"
  | "running"
  | "review"
  | "completed"
  | "failed";

export interface CampaignOutput {
  id: string;
  output_type: string;
  content: string | null;
  asset_url: string | null;
  metadata: Record<string, unknown>;
  created_at: string;
}

export interface Campaign {
  id: string;
  user_id: string;
  project_id?: string | null;
  title: string;
  brief: string;
  status: CampaignStatus;
  state: Record<string, unknown>;
  error_message: string | null;
  created_at: string;
  updated_at: string;
  outputs: CampaignOutput[];
}

export interface CreateCampaignPayload {
  user_id: string;
  title: string;
  brief: string;
  brand_voice?: string;
  target_audience?: string;
}

export interface CampaignSummary {
  id: string;
  project_id?: string | null;
  title: string;
  status: CampaignStatus;
  task_type?: TaskType | null;
  created_at: string;
  updated_at: string;
}

File: backend/app/schemas/campaign.py
import uuid
from datetime import datetime

from pydantic import BaseModel, Field

from app.db.models import CampaignStatus


class CampaignCreate(BaseModel):
    user_id: str = Field(..., min_length=1, max_length=128)
    title: str = Field(..., min_length=1, max_length=256)
    brief: str = Field(..., min_length=10)
    project_id: uuid.UUID | None = None
    brand_voice: str | None = None
    target_audience: str | None = None


class CampaignOutputRead(BaseModel):
    id: uuid.UUID
    output_type: str
    content: str | None
    asset_url: str | None
    metadata: dict = Field(default_factory=dict, validation_alias="metadata_")
    created_at: datetime

    model_config = {"from_attributes": True}


class CampaignRead(BaseModel):
    id: uuid.UUID
    user_id: str
    project_id: uuid.UUID | None = None
    title: str
    brief: str
    status: CampaignStatus
    state: dict
    error_message: str | None
    created_at: datetime
    updated_at: datetime
    outputs: list[CampaignOutputRead] = []

    model_config = {"from_attributes": True}


class CampaignRunResponse(BaseModel):
    campaign_id: uuid.UUID
    status: CampaignStatus
    job_id: str | None = None


class CampaignSummaryRead(BaseModel):
    id: uuid.UUID
    project_id: uuid.UUID | None = None
    title: str
    status: CampaignStatus
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}
